### PR TITLE
Serial

### PR DESCRIPTION
**Pull request by RJ-Russell**
Can configure Usart for serial transmission. Implemented print, println, kprint, and kprintln functionality. 

@pdouglas94 
@dnseitz 
@benrhiger 
@tgjamin 
@moinoue 
@L0ntra

### CODE REVIEW

File: altos-rust/Makefile [hunk 1]
```diff
@@ -90,11 +90,11 @@ test:
 
 doc:
 	@mkdir -p $(build_path)
-	@$(cargo) doc $(doc_args)
-	@cp -r $(doc_target_path) $(doc_build_path)
+	@cd port/cortex-m0; \
```
Comment (pdouglas94) [inline altos-rust/Makefile:93]: I don't see any docs being generated for the usart directory when i run `make doc`. I think it has something to do with the conditional compilation of your serial code.

File: altos-rust/altos-core/src/queue/ringbuffer.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    pub fn is_empty(&self) -> bool {
+        self.start == self.end && !self.full
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+#[test]
```
Comment (pdouglas94) [inline altos-rust/altos-core/src/queue/ringbuffer.rs]: #[test] Should be aligned with the indentation of the functions.

Comment (pdouglas94) [reply altos-rust/altos-core/src/queue/ringbuffer.rs]: This applies to the functions below as well.

File: altos-rust/altos-core/src/sched/mod.rs [hunk 3]
```diff
@@ -96,6 +96,7 @@ pub fn switch_context() {
 
 /// Start running the first task in the queue
 pub fn start_scheduler() {
+    kprintln!("Starting Scheduler");
```
Comment (pdouglas94) [inline altos-rust/altos-core/src/sched/mod.rs]: We probably shouldn't leave print statements in.

File: altos-rust/port/cortex-m0/src/lib.rs [hunk 4]
```diff
@@ -41,172 +44,184 @@ pub use exceptions::EXCEPTIONS;
...
-      "d_done:\n")
-    : /* no outputs */
-    : /* no inputs */
-    : "r0", "r1", "r2", "r3" /* clobbers */
-    : "volatile");
-  }
+#[cfg(target_arch="arm")]
+    unsafe {
+        asm!(
+                concat!(
```
Comment (pdouglas94) [inline altos-rust/port/cortex-m0/src/lib.rs]: Why is everything inside asm! indented twice?

Comment (RJ-Russell) [reply altos-rust/port/cortex-m0/src/lib.rs]: because indentation is like cowbell.

Comment (RJ-Russell) [inline altos-rust/port/cortex-m0/src/lib.rs]: and actually its not.. that is the correct indentation.

Comment (dnseitz) [reply altos-rust/port/cortex-m0/src/lib.rs]: All the strings are technically inside the `concat!` macro, this just makes it easier to comment lines

Comment (dnseitz) [reply altos-rust/port/cortex-m0/src/lib.rs]: Oh, I see what you're saying, yeah... nice job RJ...

**Review by pdouglas94 — APPROVED** [review]

**Pull request merged**